pkg/api/httpcontext/httpcontext.go

```diff
@@ -0,0 +1,76 @@
+package httpcontext
+
+import (
+	"math/rand"
+	"net/http"
+	"strconv"
+	"sync"
+
+	"github.com/golang/glog"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+)
+
+const header = "X-Trace"
+
+var (
+	lock     sync.RWMutex
+	requests map[*http.Request]api.Context
+)
+
+func init() {
+	requests = make(map[*http.Request]api.Context)
+}
+
+func For(req *http.Request) api.Context {
+	lock.RLock()
+	defer lock.RUnlock()
+	ctx, ok := requests[req]
```

Comment:
Suggestion: instead of using a pointer as a key, actually modify the request, adding an X-Trace header if one isn't found. Then make the map from X-Trace to ctx instead of from pointer to ctx.

Using pointers breaks as soon as someone copies a request... not that people should do that, but it feels kinda bad.